File: 한민선/240521/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        Queue<Integer> q = new LinkedList<>();
        Stack<Integer> stack = new Stack<>();

        int n = Integer.parseInt(br.readLine());
        for (int i = 0; i < n; i++) {
            q.add(Integer.parseInt(br.readLine()));
        }

        for (int i = 1; i < n+1; i++) {
            if(!stack.isEmpty()){
                while(!stack.isEmpty()){
                    int num = stack.pop();
                    if(num == q.peek()){
                        q.poll();
                        sb.append("- \n");
                    }else{
                        stack.push(num);
                        break;
                    }
                }
            }
            stack.push(i);
            sb.append( "+ \n");
        }

        int c = stack.size();
        for (int i = 0; i < c; i++) {
            int num = stack.pop();
            if(num == q.peek()){
                q.poll();
                sb.append("- \n");
            }else{
                stack.push(num);
            }
        }

        if(!stack.isEmpty()){
            System.out.println("NO");
        }else{
            System.out.println(sb);
        }
    }
}

File: 이하영/240523/1874. 스택 수열/스택 수열.py
#pop을 해서 입력한 수열을 완성시키면 됨
#단, 스택에는 오름차순대로 한 번씩 push할 수 있음. 
#ex) 1, 2, 3, 4 push 후 4,3 pop 다시 5, 6 push 후 6 pop

# 1. 수열 입력 받기
# 2. 스택에 해당 수열의 값만큼 오름차 순대로 push, op에 + 저장
# 3. 스택의 맨 위 값이 해당 수열의 값과 같으면 pop, op에 - 저장
# 3-1. 같지 않다면 no 출력

import sys

n = int(sys.stdin.readline())
stack = [] 
op = [] # +, - 저장
count = 0
possible = True

for _ in range(n):
  num = int(sys.stdin.readline())

  while num > count:
    count += 1
    stack.append(count)
    op.append('+')

  if stack[-1] == num:
    stack.pop()
    op.append('-')
  else:
    possible = False
    break

if possible:
  for i in op:
    print(i)
else:
  print("NO")